auth/src/main/java/io/grpc/auth/GoogleAuthLibraryCallCredentials.java

```diff
@@ -245,31 +244,59 @@ static JwtHelper createJwtHelperOrNull(ClassLoader loader) {
     return rawGoogleCredentialsClass.asSubclass(Credentials.class);
   }
 
+  private static class MethodPair {
+    private final Method getter;
+    private final Method builderSetter;
+
+    private MethodPair(Method getter, Method builderSetter) {
+      this.getter = getter;
+      this.builderSetter = builderSetter;
+    }
+
+    private void apply(Credentials credentials, Object builder)
+        throws InvocationTargetException, IllegalAccessException {
+      builderSetter.invoke(builder, getter.invoke(credentials));
+    }
+  }
+
   @VisibleForTesting
   static class JwtHelper {
     private final Class<? extends Credentials> serviceAccountClass;
-    private final Constructor<? extends Credentials> jwtConstructor;
+    private final Method newJwtBuilder;
+    private final Method build;
     private final Method getScopes;
-    private final Method getClientId;
-    private final Method getClientEmail;
-    private final Method getPrivateKey;
-    private final Method getPrivateKeyId;
+    private final List<MethodPair> methodPairs;
 
     public JwtHelper(Class<?> rawServiceAccountClass, ClassLoader loader)
         throws ClassNotFoundException, NoSuchMethodException {
       serviceAccountClass = rawServiceAccountClass.asSubclass(Credentials.class);
       getScopes = serviceAccountClass.getMethod("getScopes");
-      getClientId = serviceAccountClass.getMethod("getClientId");
-      getClientEmail = serviceAccountClass.getMethod("getClientEmail");
-      getPrivateKey = serviceAccountClass.getMethod("getPrivateKey");
-      getPrivateKeyId = serviceAccountClass.getMethod("getPrivateKeyId");
       Class<? extends Credentials> jwtClass = Class.forName(
           "com.google.auth.oauth2.ServiceAccountJwtAccessCredentials", false, loader)
           .asSubclass(Credentials.class);
-      jwtConstructor
-          = jwtClass.getConstructor(String.class, String.class, PrivateKey.class, String.class);
+      newJwtBuilder = jwtClass.getDeclaredMethod("newBuilder");
+      Class<?> builderClass = newJwtBuilder.getReturnType();
+      build = builderClass.getMethod("build");
+
+      methodPairs = new ArrayList<>();
+      for (String key : new String[] { "ClientId", "ClientEmail", "PrivateKey", "PrivateKey" }) {
+        Method getter = serviceAccountClass.getMethod("get" + key);
```

Comment:
The string generation here a bit worrying, since things like ProGuard can rename methods, and they won't be able to notice references like this. We've had to be explicit and repeat ourselves many times in the past. That generally means the string arguments have to be literals.

```java
methodPairs.add(new MethodPair(
    serviceAccountClass.getMethod("getClientId"),
    builderClass.getMethod("setClientId", getter.getReturnType())));
```

Reply:
Will do.  Thanks for the catch.